### DISCUSSION TITLE
Run 'onchange_after_' script based on expanded template contents

### DISCUSSION THREAD

**Discussion by woolsweater** [twpayne/chezmoi#2684] [Q&A] [answered]
I would like to [use the tip from the user guide about triggering an `onchange_after_` script with another file](https://www.chezmoi.io/user-guide/use-scripts-to-perform-actions/#run-a-script-when-the-contents-of-another-file-changes), but with an _expanded template_ as the trigger. In my case the template contents are shell code like the script, so it's problematic to include them directly with `{{ template "foo" . }}`. Unfortunately there doesn't seem to be any other way to do it that I can find.

---

I have tried a few different things to hack this together:

The most obvious solution, moving the template contents into `.chezmoitemplates/foo`, and putting `{{ template "foo" . | sha256sum }}` into the script doesn't work because [apparently the `template` action doesn't produce a string that can be passed to a function](https://stackoverflow.com/questions/69662568/go-template-post-processing-is-it-possible?noredirect=1&lq=1).

In the script `{{ include (joinPath .chezmoi.homeDir ".foo") | sha256sum  }}` isn't effective because, when calculating the script's state, it uses the _current_ contents of '.foo' (before chezmoi applies anything). I would have to run `chezmoi apply` twice in a row to make the trigger work.

Adding another layer of indirection: `.chezmoitemplates/foo` then `.chezmoitemplates/foo-expanded` that just contains `{{ template "foo" . }}`. The script then has `{{ include ".chezmoitemplates/foo-expanded" | sha256sum }}` but this of course includes the literal contents rather than post-expansion.

---

Did I miss a way to make this work? If not, would that be a desireable feature for chezmoi?

I don't know Go or chezmoi's internals but the most straightforward solution appears to be a template helper function that can render a template to a string, so it can be passed to other functions: `{{ templateAsString "foo" . | sha256sum }}`.

**Comment by twpayne** [comment, marked as answer]
Have you tried the [`includeTemplate` function](https://www.chezmoi.io/reference/templates/functions/includeTemplate/)?

**Reply by woolsweater** [reply]
Brilliant, that's what I missed! Thanks!
